source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.h

```diff
@@ -20,17 +20,52 @@ namespace HeaderToMetadataFilter {
 using ProtoRule = envoy::extensions::filters::http::header_to_metadata::v3::Config::Rule;
 using ValueType = envoy::extensions::filters::http::header_to_metadata::v3::Config::ValueType;
 using ValueEncode = envoy::extensions::filters::http::header_to_metadata::v3::Config::ValueEncode;
+using KeyValuePair = envoy::extensions::filters::http::header_to_metadata::v3::Config::KeyValuePair;
+
+// Interface for getting values.
+class ValueSelector {
+public:
+  virtual absl::optional<std::string> extract(Http::HeaderMap& map) const PURE;
+  virtual std::string toString() const PURE;
+  virtual ~ValueSelector() = default;
+};
+
+// Get values from header.
+class HeaderValueSelector : public ValueSelector
+{
+public:
+  ~HeaderValueSelector() override = default;
+  explicit HeaderValueSelector(Http::LowerCaseString header, bool remove) : header_(std::move(header)), remove_(std::move(remove)) {}
+  absl::optional<std::string> extract(Http::HeaderMap& map) const override;
+  std::string toString() const override { return fmt::format("header '{}'", header_.get()); }
+
+private:
+  const Http::LowerCaseString header_;
+  const bool remove_;
+};
+
+// Get values from cookie.
+class CookieValueSelector : public ValueSelector
+{
+public:
+  ~CookieValueSelector() override = default;
+  explicit CookieValueSelector(std::string cookie) : cookie_(std::move(cookie)) {}
+  absl::optional<std::string> extract(Http::HeaderMap& map) const override;
+  std::string toString() const override { return fmt::format("cookie '{}'", cookie_); }
+
+private:
+  const std::string cookie_;
+};
 
 class Rule {
 public:
-  Rule(const std::string& header, const ProtoRule& rule);
+  Rule(const ProtoRule& rule);
   const ProtoRule& rule() const { return rule_; }
   const Regex::CompiledMatcherPtr& regexRewrite() const { return regex_rewrite_; }
   const std::string& regexSubstitution() const { return regex_rewrite_substitution_; }
-  const Http::LowerCaseString& header() const { return header_; }
+  std::shared_ptr<const ValueSelector> vs;
```

Comment:
Envoy style it to append `_` to field names. We'd prefer a more descriptive name as well.